File: forgerock-auth/src/main/java/org/forgerock/android/auth/PolicyAdvice.java
/*
 * Copyright (c) 2019 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth;

import androidx.annotation.NonNull;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import java.io.Serializable;
import java.io.StringReader;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;

import lombok.Builder;
import lombok.Getter;

/**
 * Domain object for advice. The Advice is received from resource API for Step up authentication
 */
@Builder
@Getter
public class PolicyAdvice implements Serializable {

    public static final int AUTH_LEVEL_CONDITION_ADVICE = 1;
    public static final int AUTH_SCHEME_CONDITION_ADVICE = 2;
    public static final int AUTHENTICATE_TO_REALM_CONDITION_ADVICE = 3;
    public static final int AUTHENTICATE_TO_SERVICE_CONDITION_ADVICE = 4;
    public static final int SESSION_CONDITION_ADVICE = 5;
    public static final int TRANSACTION_CONDITION_ADVICE = 6;
    public static final int UNKNOWN = -1;

    @lombok.NonNull
    private String type;
    @lombok.NonNull
    private String value;

    @NonNull
    @Override
    public String toString() {
        return "<Advices>" +
                "<AttributeValuePair>" +
                "<Attribute name=\"" + type + "\"/>" +
                "<Value>" + value + "</Value>" +
                "</AttributeValuePair>" +
                "</Advices>";
    }

    public int getType() {
        switch (type) {
            case "AuthenticateToServiceConditionAdvice":
                return AUTHENTICATE_TO_SERVICE_CONDITION_ADVICE;
            case "TransactionConditionAdvice":
                return TRANSACTION_CONDITION_ADVICE;
            default:
                return UNKNOWN;
        }
    }


    /**
     * Parse the advice xml
     *
     * @param xml The Advice in xml form
     * @return The parsed Policy Advice
     */
    public static PolicyAdvice parse(String xml) {
        try {
            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
            DocumentBuilder builder = factory.newDocumentBuilder();
            Document document = builder.parse(new InputSource(new StringReader(xml)));
            NodeList attributes = document.getElementsByTagName("AttributeValuePair");

            for (int i = 0; i < attributes.getLength(); i++) {

                Node node = attributes.item(i);
                if (node.getNodeType() == Node.ELEMENT_NODE) {
                    Element element2 = (Element) node;
                    String attribute = element2.getElementsByTagName("Attribute").item(0).getAttributes().getNamedItem("name").getNodeValue();
                    String value = element2.getElementsByTagName("Value").item(0).getFirstChild().getNodeValue();
                    return PolicyAdvice.builder()
                            .type(attribute)
                            .value(value).build();
                }
            }
            throw new IllegalArgumentException("Failed to parse policy advice");
        } catch (Exception e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: forgerock-auth/src/main/java/org/forgerock/android/auth/PolicyAdvice.kt
/*
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 *  This software may be modified and distributed under the terms
 *  of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth

import android.util.Base64
import org.json.JSONException
import org.json.JSONObject
import org.w3c.dom.Element
import org.w3c.dom.Node
import org.xml.sax.InputSource
import java.io.Serializable
import java.io.StringReader
import java.nio.charset.StandardCharsets
import javax.xml.parsers.DocumentBuilderFactory

/**
 * Domain object for advice. The Advice is received from resource API for Step up authentication
 */
class PolicyAdvice internal constructor(val type: String, val value: String) :
    Serializable {

    override fun toString(): String {
        return "<Advices>" +
                "<AttributeValuePair>" +
                "<Attribute name=\"" + type + "\"/>" +
                "<Value>" + value + "</Value>" +
                "</AttributeValuePair>" +
                "</Advices>"
    }

    fun getType(): Int {
        return when (type) {
            "AuthenticateToServiceConditionAdvice" -> AUTHENTICATE_TO_SERVICE_CONDITION_ADVICE
            "TransactionConditionAdvice" -> TRANSACTION_CONDITION_ADVICE
            else -> UNKNOWN
        }
    }

    class PolicyAdviceBuilder internal constructor() {
        private var type: String = ""
        private var value: String = ""
        fun type(type: String): PolicyAdviceBuilder {
            this.type = type
            return this
        }

        fun value(value: String): PolicyAdviceBuilder {
            this.value = value
            return this
        }

        fun build(): PolicyAdvice {
            return PolicyAdvice(type, value)
        }

        override fun toString(): String {
            return "PolicyAdvice.PolicyAdviceBuilder(type=$type, value=$value)"
        }
    }

    companion object {
        const val AUTH_LEVEL_CONDITION_ADVICE = 1
        const val AUTH_SCHEME_CONDITION_ADVICE = 2
        const val AUTHENTICATE_TO_REALM_CONDITION_ADVICE = 3
        const val AUTHENTICATE_TO_SERVICE_CONDITION_ADVICE = 4
        const val SESSION_CONDITION_ADVICE = 5
        const val TRANSACTION_CONDITION_ADVICE = 6
        const val UNKNOWN = -1

        @JvmStatic
        fun builder(): PolicyAdviceBuilder {
            return PolicyAdviceBuilder()
        }

        /**
         * Parse the advice xml
         *
         * @param advice The Advice in xml form
         * @return The parsed Policy Advice
         */
        @JvmStatic
        fun parse(advice: String): PolicyAdvice {
            val factory = DocumentBuilderFactory.newInstance()
            val builder = factory.newDocumentBuilder()
            val document = builder.parse(InputSource(StringReader(advice)))
            val attributes = document.getElementsByTagName("AttributeValuePair")
            for (i in 0 until attributes.length) {
                val node = attributes.item(i)
                if (node.nodeType == Node.ELEMENT_NODE) {
                    val element2 = node as Element
                    val attribute = element2.getElementsByTagName("Attribute")
                        .item(0).attributes.getNamedItem("name").nodeValue
                    val value =
                        element2.getElementsByTagName("Value").item(0).firstChild.nodeValue
                    return builder()
                        .type(attribute)
                        .value(value).build()
                }
            }
            throw IllegalArgumentException("Advice Not Found: $advice")
        }

        /**
         * Parse the advice as base64 encoded xml String
         * @param Base64 encoded xml String
         * @return The parsed Policy Advice
         */
        fun parseAsBase64XML(base64XML: String): PolicyAdvice {
            return parse(String(Base64.decode(base64XML,
                Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP), StandardCharsets.UTF_8))
        }


        /**
         * Parse the advice as base64 encoded json String
         * @param Base64 encoded json String
         * @return The parsed Policy Advice
         */
        fun parseAsBase64(advice: String?): PolicyAdvice {
            val jsonObject = JSONObject(String(Base64.decode(advice,
                Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)))
            return builder()
                .type("TransactionConditionAdvice")
                .value(jsonObject.getJSONArray("TransactionConditionAdvice").getString(0))
                .build()
        }
    }
}